Exports off of couch - redux

NoahCarnahan opened the pull request:
This PR un-reverts https://github.com/dimagi/commcare-hq/pull/10477, and continues on that work.
- Fixes a variety of bugs
- Implements missing views
- Adds missing export row number functionality
- Adds a `user_type` field to the xform es index

This PR is not ready to merge, just putting it up now for review. I think I want to run QA on staging, and we also need to time the xform reindex appropriately. I also need to do a few more things, like actually use the `user_type` field in the export filters.

Should be reviewed commit-by-commit. The first commit was PR https://github.com/dimagi/commcare-hq/pull/10477, so all those changes were reviewed there.

NoahCarnahan committed:
Move get_properly_wrapped_export_instance() to dbaccessors

NoahCarnahan committed:
Add view and function for retrieving all daily saved exports

NoahCarnahan committed:
Rebuild new style daily saved exports in periodic task

NoahCarnahan committed:
Fix botched keyword args in rebuild_export_task()

NoahCarnahan committed:
Add missing isLegacy flag

NoahCarnahan committed:
Remove default filter that breaks unknown users export filter

NoahCarnahan committed:
Add view for bulk new style exports

NoahCarnahan committed:
Finish removing output_dir from export functions

NoahCarnahan committed:
Revert "Revert "Get daily saved exports off the couch""

This reverts commit 5782207c356bdab773becb94caa7e3e08ab21efd.

Conflicts:
	corehq/apps/export/views.py

NoahCarnahan committed:
Fix FilterCaseESExportDownloadForm return type

NoahCarnahan committed:
Use correct download URL in export list

NoahCarnahan committed:
Add view for downloading new style case exports

NoahCarnahan committed:
Remove misc todos

NoahCarnahan committed:
Fix sidebars in new export views

NoahCarnahan committed:
Fix new style bulk export file name

NoahCarnahan committed:
Update es xform mapping

Adds a null_value to form.meta.userID.
Also updates ESQuery to automatically convert None to this null value
This change will fix the new export "unknown users" filter

NoahCarnahan committed:
Add support for row number columns in exports

NoahCarnahan committed:
Store user type in es

dimagimon commented:
There are 7 errors:

* corehq/apps/export/models/new.py, line 255 - E501 line too long (131 > 115 characters)
* corehq/apps/export/tests/test_daily_saved_exports.py, line 7 - E302 expected 2 blank lines, found 1
* corehq/apps/export/tests/test_daily_saved_exports.py, line 29 - E301 expected 1 blank line, found 0
* corehq/apps/export/views.py, line 1557 - E303 too many blank lines (3)
* corehq/apps/reports/urls.py, line 88 - E501 line too long (142 > 115 characters)
* corehq/pillows/mappings/xform_mapping.py, line 6 - E303 too many blank lines (3)
* corehq/couchapps/export_instances_by_is_daily_saved/views/view/map.js, line 1 - Missing name in function declaration.

NoahCarnahan committed:
[ci-skip] lint

dimagimon commented on corehq/apps/export/models/new.py:
```diff
@@ -222,14 +241,18 @@ def get_column(self, item_path):
+        >>> TableConfiguration._get_sub_documents_helper(['foo'], [{'foo': {'bar': 'a'}}, {'foo': {'bar': 'b'}}])
         [{'bar': 'a'}, {'bar': 'b'}]
-        >>> TableConfiguration()._get_sub_documents(['foo', 'bar'], [{'foo': [{'bar': {'baz': 'a'}}, {'bar': {'baz': 'b'}},]}]
+        >>> TableConfiguration._get_sub_documents_helper(['foo', 'bar'], [{'foo': [{'bar': {'baz': 'a'}}, {'bar': {'baz': 'b'}},]}]
```
E501 line too long (131 > 115 characters)

dimagimon replied:
E501 line too long (131 > 115 characters)

dimagimon commented on corehq/apps/reports/urls.py:
```diff
@@ -85,6 +85,7 @@
     url(r"^export/bulk/download/$", "export_default_or_custom_data", name="export_bulk_download", kwargs=dict(bulk_export=True)),
     # saved
     url(r"^export/saved/download/(?P<export_id>[\w\-]+)/$", "hq_download_saved_export", name="hq_download_saved_export"),
+    url(r"^export/new/saved/download/(?P<export_instance_id>[\w\-]+)/$", "hq_download_new_saved_export", name="hq_download_new_saved_export"),
```
E501 line too long (142 > 115 characters)

dimagimon commented on corehq/couchapps/export_instances_by_is_daily_saved/views/view/map.js:
```diff
@@ -0,0 +1,5 @@
+function(doc) {
```
Missing name in function declaration.

dimagimon replied:
Missing name in function declaration.

dimagimon replied:
Missing name in function declaration.

dimagimon replied:
Missing name in function declaration.

dimagimon replied:
Missing name in function declaration.

dimagimon replied:
Missing name in function declaration.

dimagimon replied:
Missing name in function declaration.

dimagimon replied:
Missing name in function declaration.

dimagimon replied:
Missing name in function declaration.

benrudolph commented:
sweet this is looking real nice

NoahCarnahan committed:
Improve _get_user_type()

NoahCarnahan committed:
Fix pillow test

snopoke commented:
not mergable

NoahCarnahan committed:
Add user type fields to case and form elasticsearch documents

Also adds
- case owner_type es filter
- forms user_type es filter
- remove todo
- OwnerTypeFilter, UserTypeFilter export filters
- update export forms to use UserTypeFilter and OwnerTypeFilter

NoahCarnahan committed:
Fix form es user_id filter

NoahCarnahan committed:
Merge remote-tracking branch 'origin/master' into sql-saved-exports-2

Conflicts:
	testapps/test_pillowtop/tests/data/all-pillow-meta.json

NoahCarnahan committed:
Fix mocha tests

NoahCarnahan committed:
Fix pillow tests maybe

NoahCarnahan committed:
[ci-skip] update doc string

NoahCarnahan committed:
Fix API test

NoahCarnahan committed:
Remove no longer relevant test

NoahCarnahan committed:
Create version of elasticsearch.helpers.scan that returns the total number of hits

dimagimon commented on corehq/pillows/case.py:
```diff
@@ -104,3 +112,21 @@ def get_couch_case_reindexer():
 def get_sql_case_reindexer():
     return PillowReindexer(get_sql_case_to_elasticsearch_pillow(), SqlCaseChangeProvider())
+
+@quickcache(['domain', 'user_id'], timeout=ONE_DAY)
```
E302 expected 2 blank lines, found 1

dimagimon commented on corehq/pillows/case.py:
```diff
@@ -104,3 +112,21 @@ def get_couch_case_reindexer():
+    Get the username for the given user_id. We are replicating the beahvior of
+    corehq.apps.reports.util.get_all_users_by_domain
+    """
+    # TODO: This assumes that the form index will be built before the case index. Will that always be true? Is this safe?
```
E501 line too long (121 > 115 characters)

NoahCarnahan committed:
Eliminate the admin user export filter

NoahCarnahan committed:
Revert "Fix API test"

This reverts commit a56c0861454e1b232557fcbf8b275a0d49c86d52.

NoahCarnahan committed:
Track download progress

NoahCarnahan committed:
Fix pillow instantiation test

dimagimon commented on corehq/pillows/utils.py:
```diff
@@ -1,4 +1,23 @@
+from corehq.apps.commtrack.const import COMMTRACK_USERNAME
+from corehq.apps.reports.models import HQUserType
```
F401 'HQUserType' imported but unused

NoahCarnahan committed:
[ci skip] lint

NoahCarnahan committed:
Fetch export docs in smaller chunks than the default 1 million

NoahCarnahan commented:
@dimagi/scale-team this is ready for review again.
The primary changes since opening the PR are:
- Add user type to form and case es indexes
- eliminate the "Admin" export filter
- track download progress


I think we're very close to being able to deploy. The two things left on my radar are:
- Implement the special system properties
- Finish off the functions that migrate existing exports

My plan is to deploy this branch to staging if the travis build passes

benrudolph commented:
this looks good to me :+1: ping when you think you've done enough testing on staging. would be good to and do the other two things on your radar in a separate PR

NoahCarnahan committed:
Fix testReportXFormProcessing

NoahCarnahan committed:
Merge branch 'master' into sql-saved-exports-2

Conflicts:
	corehq/apps/es/forms.py

benrudolph commented:
dang we were supposed to merge this on friday weren't we

NoahCarnahan commented:
Yeah, sorry about that. Let's discuss the best approach when you come in.

NoahCarnahan committed:
Merge remote-tracking branch 'origin/master' into sql-saved-exports-2

NoahCarnahan committed:
Revert es index changes

I'll put these changes into a different branch to PR separately.

dimagimon commented on corehq/apps/export/tests/test_export_filters.py:
```diff
@@ -11,15 +11,26 @@
 from corehq.apps.export.models.new import (
     CaseExportInstance,
+    FormExportInstance)
+from corehq.apps.export.tests.util import (
```
F401 'DEFAULT_APP_ID' imported but unused

dimagimon replied:
F401 'DEFAULT_XMLNS' imported but unused

NoahCarnahan commented:
@benrudolph Alright I've removed the es index changes from this PR, so it should be good to merge now. I left in the new functions like `FormES.user_type()`, which require reindex to work for simplicities sake, but happy to remove them if you think it could cause confusion to keep them.

benrudolph commented:
yeeha :horse:! nope i think it's good to leave

benrudolph commented:
:fireworks: :fireworks: :fireworks:

czue commented:
:balloon:

NoahCarnahan commented:
:joy:
